src/source/Ice/IceAgent.c

```diff
@@ -1460,69 +1436,15 @@ STATUS iceAgentInitRelayCandidate(PIceAgent pIceAgent)
-                                                  &pNewCandidate->pSocketConnection));
-                // connectionListener will free the pSocketConnection at the end.
-                CHK_STATUS(connectionListenerAddConnection(pIceAgent->pConnectionListener,
-                                                           pNewCandidate->pSocketConnection));
-
-                pNewCandidate->iceCandidateType = ICE_CANDIDATE_TYPE_RELAYED;
-                pNewCandidate->state = ICE_CANDIDATE_STATE_NEW;
-                pNewCandidate->iceServerIndex = j;
-                pNewCandidate->foundation = pIceAgent->foundationCounter++; // we dont generate candidates that have the same foundation.
-                pNewCandidate->priority = computeCandidatePriority(pNewCandidate);
-
-                CHK_STATUS(createTurnConnection(&pIceAgent->iceServers[j], pIceAgent->timerQueueHandle, TURN_CONNECTION_DATA_TRANSFER_MODE_SEND_INDIDATION,
-                                                KVS_ICE_DEFAULT_TURN_PROTOCOL, NULL, pNewCandidate->pSocketConnection, pIceAgent->pConnectionListener,
-                                                &pCurrentTurnConnectionTracker->pTurnConnection));
-                pCurrentTurnConnectionTracker->pRelayCandidate = pNewCandidate;
-                pCurrentTurnConnectionTracker->pIceAgent = pIceAgent;
-                pCurrentTurnConnectionTracker->freeTurnConnectionTimerId = UINT32_MAX;
-                pCurrentTurnConnectionTracker->freeTurnConnectionEndTime = INVALID_TIMESTAMP_VALUE;
-                pNewCandidate->pTurnConnectionTracker = pCurrentTurnConnectionTracker;
-
-                CHK_STATUS(doubleListInsertItemHead(pIceAgent->localCandidates, (UINT64) pNewCandidate));
-                pNewCandidate = NULL;
-
-                MUTEX_LOCK(pIceAgent->lock);
-                locked = TRUE;
-
-                /* add existing remote candidates to turn. Need to acquire lock because remoteCandidates can be mutated by
-                 * iceAgentAddRemoteCandidate calls. */
-                CHK_STATUS(doubleListGetHeadNode(pIceAgent->remoteCandidates, &pCurNode));
-                while (pCurNode != NULL) {
-                    CHK_STATUS(doubleListGetNodeData(pCurNode, &data));
-                    pCurNode = pCurNode->pNext;
-                    pCandidate = (PIceCandidate) data;
-
-                    // TODO: Stop skipping IPv6. Since we're allowing IPv6 remote candidates from iceAgentAddRemoteCandidate for host candidates,
-                    // it's possible to have a situation where the turn server uses IPv4 and the remote candidate uses IPv6.
-                    if (IS_IPV4_ADDR(&pCandidate->ipAddress)) {
-                      CHK_STATUS(turnConnectionAddPeer(pCurrentTurnConnectionTracker->pTurnConnection,
-                                                       &pCandidate->ipAddress));
-                    }
+                if (pIceAgent->iceServers[j].transport == KVS_SOCKET_PROTOCOL_UDP ||
```

Comment:
Perhaps instead of the else cases you could simply pass the pIceAgent->iceServers[j].transport param directly to iceAgentInitRelayCandidate and handle the NONE case there?

Reply:
I cant, each turnConnection has to be either UDP to TCP. turnConnection cant create two copies of itself for UDP and TCP, has to be done by the caller.

Reply:
What I mean is in else clause here to call CHK_STATUS(iceAgentInitRelayCandidate(pIceAgent, pSocketAddrForTurn, j, pIceAgent->iceServers[j].transport)); directly.

then, inside the iceAgentInitRelayCandidate, do the assumption that if the transport is NONE then assume it UDP - like what you are doing here. 

In general, having this else if/else if with both if statements ORing with PROTOCOL_NONE is not a good idea as both with match the NONE but the first will take precedence.

Reply:
what I want to do here is if transport is NONE, then create both TCP and UDP. if I pass transport to iceAgentInitRelayCandidate and assume UDP if it is NONE, it will create two UDPs which is not what I want.

Reply:
Oh, got it. Didn't notice that the second one is not else if